### DISCUSSION TITLE
Set up a volume using multiple domains without @web

### DISCUSSION THREAD

**Discussion by tmrkp** [craftcms/cms#10445] [Q&A]
I'm trying to configure a local volume for multiple sites that each have their own domain. My goal is to use the same asset on different sites. But the links to the asset in the front-end should always use the sites domain in order to avoid cross-origin urls.

I've configured the base urls for the sites using environment variables, but for the volume base url this is not possible since the domain is not the same for different sites.

I could only achieve my goal by using `@web` in the volume base url. But this seems to be discouraged.

What is the correct way to set this up?

**Comment by andris-sevcenko** [comment]
As a quick hack, you can use `/<path>` as the Volume URL and just skip the domain.

**Reply by andris-sevcenko** [reply]
Disclaimer: this is probably also discouraged and incredibly volatile.

**Reply by mmikkel** [reply]
Using the `@web` alias is only discouraged if it's not explicitly set, right? So overriding it in `config/general.php` like below should be considered a safe workaround for OP?

```php
'aliases' => [
    '@web' => App::env('PRIMARY_SITE_URL'),
];
```

**Reply by tmrkp** [reply]
I've tried just using `/<path>` and it works, but I also feel like it would be bad practive. Using assets in email templates and other places outsite the website won't work properly.

Setting the `@web` alias means assets are always linked with the same domain on all sites. I'd like to avoid that, because I don't want content to be linked to a different domain.

**Reply by timkelty** [reply]
@timrosskamp I would set up custom aliases for each site and use those, eg:

```
'aliases' => [
    '@webDe' => App::env('SITE_URL_DE'),
    '@webEn' => App::env('SITE_URL_EN'),
    '@webGb' => App::env('SITE_URL_GB'),
];
```

**Reply by tmrkp** [reply]
@timkelty But I can only assign one of them to one asset volume base path. Which would then set that domain on assets for all the sites. What I'd need is to set the `@web` dynamically based on the site thats currently visited.

**Reply by timkelty** [reply]
ah! sorry, that makes sense.

> Setting the @web alias means assets are always linked with the same domain on all sites. I'd like to avoid that, because I don't want content to be linked to a different domain.

This could be a good use case for setting up a shared, `assets.domain.com`, but that it sounds like that's not what you're after.

I'd suggest to _do_ use `@web`, but set it manually. As long as you aren't using incoming `$_SERVER` values directly, you are safe.
Here's an example config:

```php
<?php
$host = getenv('HTTP_HOST');
$siteUrls = [];
switch ($host) {
    case 'site-a.com':
        define('CRAFT_SITE', 'siteA');
        $siteUrls[CRAFT_SITE] = 'https://site-a.com';
        break;
    case 'site-b.com':
      define('CRAFT_SITE', 'siteB');
      $siteUrls[CRAFT_SITE] = 'https://site-b.com';
      break;
}

return [
  'aliases' => [
      '@web' => $siteUrls[CRAFT_SITE],
  ],
  'siteUrl' => $siteUrls,
];
```

**Reply by wmdhosting** [reply]
we have done 130 site setup .    need to change   / and on every update
/vendor/craftcms/cms/src/elements

Tag.php 
Assets.php   

 isLocalized()   to false...

**Reply by timkelty** [reply]
@wmdhosting That's why I'm suggesting you use `@web`, but set it dynamically (just not directly from `$_SERVER`).
